src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -0,0 +1,516 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.CSharp;
+using static System.String;
+
+namespace CSharpSyntaxGenerator.Grammar
+{
+    internal class GrammarGenerator
+    {
+        private readonly ImmutableDictionary<string, TreeType> _nameToElement;
+
+        public GrammarGenerator(Tree tree)
+        {
+            // Syntax refers to a special pseudo-element 'Modifier'.  Just synthesize that since
+            // it's useful in the g4 grammar.
+            tree.Types.Add(new Node
+            {
+                Name = Modifier,
+                Children =
+                {
+                    new Field
+                    {
+                        Type = SyntaxToken,
+                        Kinds = Modifiers.Select(k => new Kind { Name = k.ToString() }).ToList()
+                    }
+                }
+            });
+
+            _nameToElement = tree.Types.Where(c => c is AbstractNode || c is Node)
+                                       .ToImmutableDictionary(n => n.Name);
+        }
+
+        public string Run()
+        {
+            var nodes = _nameToElement.Values;
+            var nameToProductions = nodes.ToDictionary(n => n.Name, _ => new List<Production>());
+            nameToProductions.Add(StructuredTriviaSyntax, new List<Production>());
+
+            foreach (var node in nodes)
+            {
+                if (node.Base is string base_ && base_ != CSharpSyntaxNode)
```

Comment:
nit: the underscore thing is odd. `@base`, `nodeBase`?

Reply:
sure :)